com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Utility/Imposter.hlsl

```diff
@@ -0,0 +1,393 @@
+    outUVGrid.xy = UVscaled;
+    outUVGrid.zw = grid;
+}
+void ImposterUV_oneFrame(in float3 inPos, in float4 inUV, in float imposterFrames, in float3 imposterOffset, in float imposterSize,
+    in bool isHemi, in float parallax, in int heightMapChannel, in SamplerState ss, in Texture2D heightMap,
+    out float3 outPos, out float2 outUV0, out float4 outUVGrid)
+{
+    float framesMinusOne = imposterFrames - 1;
+
+    //camera pos in object space
+    float3 objectSpaceCameraPos = mul(unity_WorldToObject, float4(_WorldSpaceCameraPos.xyz, 1)).xyz - imposterOffset;
+    float3 objectSpaceCameraDir = normalize(objectSpaceCameraPos.xyz);
+
+    //get uv in a single frame
+    float2 UVscaled = inUV.xy * (1.0 / imposterFrames);
+    float2 size = imposterSize.xx * 1.33;
+
+    float3 BillboardPos = CalculateBillboardProjection(objectSpaceCameraDir, inUV.xy);
+    BillboardPos *= 0.67;
+
+    //camera to projection vector
+    float3 rayDirLocal = BillboardPos - objectSpaceCameraPos;
+
+    //BillboardPos position to camera ray
+    float3 BillboardPosToCam = normalize(objectSpaceCameraPos - BillboardPos);
+
+    Ray ray;
+    ray.Origin = objectSpaceCameraPos;
+    ray.Direction = rayDirLocal;
+
+    //set up virtual grid
+    float2 grid;
+    if (isHemi) {
+        objectSpaceCameraDir.y = max(0.001, objectSpaceCameraDir.y);
+        grid = VectorToHemi(objectSpaceCameraDir);
+    }
+    else {
+        grid = VectorToOcta(objectSpaceCameraDir);
+    }
+
+    grid = saturate(grid * 0.5 + 0.5); //scale to 0 to 1 
+    grid *= framesMinusOne;//multiply framesMinusOne to cover the texture
+
+    float2 gridFrac = frac(grid);
+    float2 gridFloor = floor(grid);
+
+    //convert frame coord to octahedron ray
+    float3 frame0ray = normalize(GridToVector(float2(gridFloor / framesMinusOne * 2 - 1), isHemi));
+
+    float3 center = float3(0, 0, 0);
+
+    float3 plane0x;
+    float3 plane0normal = frame0ray;
+    float3 plane0z;
+    float3 frame0local = FrameTransformLocal(BillboardPosToCam, frame0ray, plane0x, plane0z);
+
+    float2 vUv0 = PlaneIntersectionUV(plane0normal, plane0x, plane0z, center, size, ray);
+    vUv0 /= imposterFrames;
+
+    frame0local.xy /= imposterFrames;
+
+    float2 gridSnap = floor(grid) / imposterFrames.xx;
+    float2 frame0grid = gridSnap;
+    float2 vp0uv = frame0grid + vUv0.xy;
+
+    if (parallax != 0) {
+        float h0 = SAMPLE_TEXTURE2D(heightMap, ss, vp0uv)[heightMapChannel];
+        vp0uv += (0.5 - h0) * frame0local.xy * parallax;
+    }
+
+    //vert pos
+    outPos = BillboardPos + imposterOffset;
+
+    //surface
+    outUV0 = vp0uv;
+    outUVGrid.xy = UVscaled;
+    outUVGrid.zw = grid;
+
+}
+//Sample from UVs 
+void ImposterSample(in float imposterFrames, in texture2D Texture, in float4 mapTexelSize, in float4 weights, in float imposterFrameClip, in float4 inUVGrid,
+    in float2 inUV0, in float2 inUV1, in float2 inUV2, in SamplerState ss, out float4 outColor)
+{
+    //frame size ->2048/12 = 170.6
```

Comment:
Can lines 342-363 be done in the ImposterUV function instead? Is there something unique about this code that makes it need to be done for each texture instead of just once? If we can move this code, we could get rid of the Frames input on this node.